### PR TITLE
Attach some tools and weapons to belt loops

### PR DESCRIPTION
**Pull request by mugling**
Previously suggested in #12089

Implements BELT_LOOP action and BELT_CLIP flag and enables the flag for some example items. Adds loops to both the firefighter and tool belts, hopefully making the former a more desireable item.

Also includes minor refactor of item properties to include accessors (fixing a TODO in the existing code).

Likely needs some scaling for balance: which belts should have loops and what should their capacity be (see max_volume and max_weight properties)

### CODE REVIEW

File: src/iuse.cpp [hunk 1]
```diff
@@ -6910,6 +6905,65 @@ int iuse::holster_ankle(player *p, item *it, bool b, const tripoint &pos)
     return it->type->charges_to_use();
 }
 
+int iuse::belt_loop (player *p, item *it, bool, const tripoint&)
+{
+    // if belt loop empty offer to attach an item otherwise detach the current one
+    if (it->contents.empty()) {
+        // display menu showing only show BELT_CLIP items
+        item *put = &(p->i_at(g->inv_for_flag("BELT_CLIP", _("Attach what to belt loop?"), false)));
+        if (put == NULL || put->is_null()) {
```
Comment (Coolthulhu) [inline src/iuse.cpp]: `put == NULL` will never be true. It can be changed to just `put->is_null()`.

File: data/json/items/armor.json [hunk 2]
```diff
@@ -6289,10 +6290,10 @@
         "id" : "fireman_belt",
         "name" : "firefighter belt",
         "volume" : 2,
-        "material_thickness" : 0,
+        "material_thickness" : 2,
```
Comment (Coolthulhu) [inline data/json/items/armor.json]: `material_thickness` is only for armor calculations. Firefighter belt has 0 coverage, meaning it will not work as armor.

File: src/iuse.cpp (hunk 1 above)
Comment (mugling) [inline src/iuse.cpp]: A similar idiom is used elsewhere in iuse.cpp. Is it appropriate to correct those as well in the same PR?

Comment (Coolthulhu) [inline src/iuse.cpp]: You can correct them there, but it's not necessary (and may slow down testing).

Also another quality thing: using references instead of pointers where possible without adding extra code. Not necessary, but makes the intent of the code clearer (pointers are only needed where they can be null or changed in the code).

Comment (mugling) [inline src/iuse.cpp]: Much of the logic is derived from code for holster/sheathes so if I update my implementation only for now and submit a later PR for updates to that code?

Comment (Coolthulhu) [inline src/iuse.cpp]: Don't submit another PR, update this one.

**Comment by BevapDin** [comment]
> Much of the logic is derived from code for holster/sheathes so if I update my implementation only for now and submit a later PR for updates to that code?

Now we have `iuse::holster_gun`, `iuse::sheath_knife`, `iuse::sheath_sword` and this new function, all with nearly the same content. At least the logic in them is very similar. Any change to one of them will often require the same change to the other ones (e.g. pointers to references), that's not good.

This code would probably work very nice as `iuse_actor`. In this case you can also get rid of the usage of *generic* and un-typed item properties and use specific, typed properties of the `iuse_actor`. Those can be validated when the item data is loaded (I'm looking at the `debugmsg( _("%s has no monster_size_capacity.")` line). This kind of error should be handled as soon as possible.

----

Another code repetition:
```C++
std::atoi(it->get_property("max_volume", "2").c_str())
```
This would be simpler for `iuse_actor` (see the existing ones), but if you want to keep the item properties, it should be part of the item functions. I suggest a system similar to the `item::get_var` functions.
```C++
long item::get_var(const std::string &name, long default_value);
int item::get_var(const std::string &name, int default_value);
double item::get_var(const std::string &name, double default_value);
std::string item::get_var(const std::string &name);
```
The type of the default value determines the type of the returned value. This concentrates all the parsing in one place.

Btw. I think writing "Get property for this item." as documentation for a function "item::get_property" is not really needed. It just repeats the function name.

**Comment by mugling** [comment]
> This code would probably work very nice as iuse_actor

iuse::sheath_knife, iuse::sheath_sword are both dependent upon flags wheres iuse::holster_gun is not. If I extend the PR to cover the former two and leave the latter for now?

> I suggest a system similar to the item::get_var functions.

Agreed, although if implement the above it won't be needed for this PR iuse::capture_monster_act() and any actions written later can use these functions

> Btw. I think writing "Get property for this item." as documentation for a function "item::get_property" is not really needed. It just repeats the function name.

Doesn't the default configuration for doxygen exclude functions without comments?

**Comment by BevapDin** [comment]
> iuse::sheath_knife, iuse::sheath_sword are both dependent upon flags wheres iuse::holster_gun is not. If I extend the PR to cover the former two and leave the latter for now?

Sure. Small PRs are easier to merge anyway.

>> I suggest a system similar to the item::get_var functions.

> Agreed, although if implement the above it won't be needed for this PR iuse::capture_monster_act() and any actions written later can use these functions

Yes, that does not need to be implemented here and now, it's unrelated, may even go into its own PR.

> Doesn't the default configuration for doxygen exclude functions without comments?

I don't know, as far as I know, everything in `/** */` is considered. I also think that not every part of the function needs to be documented separately, if the function description already says what the parameters are and what the return is, they don't have to be commented separately.

**Comment by mugling** [comment]
> Doesn't the default configuration for doxygen exclude functions without comments?
> 
> > I don't know, as far as I know, everything in /** */ is considered

Sorry the doxygen.conf sets EXTRACT_ALL=yes so such entries are indeed unnecessary

**Comment by mugling** [comment]
> If I extend the PR to cover the former two and leave the latter for now?
> 
> > Sure. Small PRs are easier to merge anyway.

Can the above refactors suffice for now and I'll look in to ```iuse_actor``` as part of a separate PR?

**Comment by BrianLefler** [comment]
How would you feel about the following tweaks, in priority order?

1. Remove max_volume, max_weight, & associated checks in iuse::belt_loop -- you already already limiting what BELT_CLIP goes on and I have a hard time visualizing why a loop cares how long an item's shaft is.  Leaving limits in seems a likely point for json misconfiguration.
2. Add BELT_CLIP to "leather belt" -- that will give back a mechanical value to that item (still worse than firefighter & tool belt.
3. (Maybe?) increase an object's encumbrance by 1 when something is in the BELT_CLIP.

File: src/iuse.cpp [hunk 3]
```diff
@@ -6910,6 +6905,65 @@ int iuse::holster_ankle(player *p, item *it, bool b, const tripoint &pos)
...
+        if (! p->is_armed() || p->wield(NULL)) {
+            item& get = it->contents[0];
+            p->inv.assign_empty_invlet(get, true);
+
+            p->moves -= get.volume() * 10;
+            p->wield(&(p->i_add(get)));
+            it->contents.erase(it->contents.begin());
+
+            p->add_msg_if_player(m_info, _("You unclip your %s from your %s!"),
+                                 get.tname().c_str(), it->tname().c_str());
```
Comment (BevapDin) [inline src/iuse.cpp]: `get` is here invalid here. It's a reference into the vector `it->contents`, but the referenced entry has just been erased (line 6957).

**Comment by mugling** [comment]
> Remove max_volume, max_weight

No, for example the firefighter belt should be able to carry a heavy or larger item than a tool belt. In a future PR I will also complete the suggestion from @BevapDin to reimplement a number of the similar iuse functions via ```iuse_actor``` and these properties would be applicable there

> Add BELT_CLIP to "leather belt"...

Yes, after the PR is complete there a number of items that might be appropriate to add but this can wait for later

> (Maybe?) increase an object's encumbrance by 1 when something is in the BELT_CLIP.

Or better still increase it by ```n``` where ```n``` scales according to the volume, weight and possibly some other values (holsters should suffer less than loops etc). When I come to look at implementing ```iuse_actor``` this could be considered.

**Comment by Coolthulhu** [comment]
That property getting part looks quite evil to me. It involves casting a string based on hardcoded argument.
I'll need someone with better code quality skills than me to comment on this.

**Comment by mugling** [comment]
@Coolthulhu Unless I misunderstand the function should be comparable to ```item::get_var``` as previously suggested?

**Comment by BevapDin** [comment]
> That property getting part looks quite evil to me. It involves casting a string based on hardcoded argument.

How so? By casting do you mean the `atof`/`atoi`/`atol` functions? They are quite save (`strtol` and similar would be better as they report errors). The same code is used in `get_var`, through a bit more verbose. I actually like the new code using the ternary better.

File: src/item.cpp [hunk 4]
```diff
@@ -2371,6 +2376,35 @@ bool item::has_flag( const std::string &f ) const
     return ret;
 }
 
+bool item::has_property (const std::string& prop) const {
+   return type->properties.find(prop) != type->properties.end();
+}
+
+std::string item::get_property (const std::string &prop, const std::string& def) const
```
Comment (kevingranade) [inline src/item.cpp]: Looks like only get_property( std::string&, std::string) and get_property(std::string, int) are used, please remove the others.
Also the API where you determine the return value based on the default value is very odd, please just distinguish between them with names, e.g. get_property_int()

File: data/json/items/armor.json (hunk 2 above)
Comment (kevingranade) [inline data/json/items/armor.json]: This comment is correct, making this attribute the equivalent of dead code, please revert it to 0.

File: src/item.cpp (hunk 4 above)
Comment (BevapDin) [inline src/item.cpp]: > Also the API where you determine the return value based on the default value is very odd, please just distinguish between them with names, e.g. get_property_int()

But why? It's simple, it's short, it's not redundant, it avoids parsing the constant default value. If given a default value of type X, the caller apparently expects to get a value of type X. Otherwise, why would the caller give a default value that would need to be converted to another type when returned.

What about a template version:
```C++
template<typename R>
R get_property(const std::string &name, R default_value = R()) {
    return ...
}

int g = get_property<int>("");
int f = get_property<int>("", 'C');
int h = get_property("", 0);
```

Comment (kevingranade) [inline src/item.cpp]: It makes perfect sense for the method to infer that it should return a type
matching a provided default value, but it makes no sense to require the
caller to provide an extraneous default value to tell the method what type
to return.

Comment (BevapDin) [inline src/item.cpp]: The default return value is always required because the property *might* not exists (what should the function return in that case?). So it's not really an extraneous parameter in my opinion.

**Comment by mugling** [comment]
> strtol and similar would be better as they report errors

Agreed. The revised implementation reports parsing errors via `debugmsg()` and returns the default value (as opposed to zero)

File: src/item.cpp [hunk 5]
```diff
@@ -2371,6 +2376,56 @@ bool item::has_flag( const std::string &f ) const
...
+double item::get_property (const std::string& prop, double def) const
+{
+    const auto it = type->properties.find(prop);
+    if (it != type->properties.end()) {
+        char  *e = nullptr;
+        double r = std::strtod(it->second.c_str(), &e);
+        if (it->second.size() && *e == '\0') {
+            return r;
+        }
+        debugmsg(_("invalid property '%s' for item '%s'"), prop.c_str(), tname().c_str());
```
Comment (BevapDin) [inline src/item.cpp]: I'm against translating debug and similar error messages (remove the `_()`). It will just lead to double translations: some player sees the translated error in their language, reports the error and translates the message back to English, which may result in different words. Good luck finding the origin of the error.

File: src/iuse.cpp [hunk 6]
```diff
@@ -6910,6 +6905,65 @@ int iuse::holster_ankle(player *p, item *it, bool b, const tripoint &pos)
...
+        // display menu showing only show BELT_CLIP items
+        item& put = p->i_at(g->inv_for_flag("BELT_CLIP", _("Attach what to belt loop?"), false));
+        if (put.is_null()) {
+            p->add_msg_if_player(_("Never mind."));
+            return 0;
+        }
+
+        // check the player selected an appropriate item
+        if (! put.has_flag("BELT_CLIP")) {
+            p->add_msg_if_player(m_info, _("You can't attach your %s to your %s!"),
```
Comment (Coolthulhu) [inline src/iuse.cpp]: Those '!'s are too "dynamic" here. '.'s would be enough.
